lensless/unrolled_admm.py

```diff
@@ -0,0 +1,266 @@
+        )
+
+    def _X_update(self, iter):
+        # to avoid computing forward model twice
+        # self._X = self._X_divmat * (self._xi + self._mu1 * self._forward_out + self._data)
+        self._X = self._X_divmat[iter] * (
+            self._xi + self._mu1[iter] * self._forward_out + self._convolver._pad(self._data)
+        )
+
+    def _image_update(self, iter):
+        rk = (
+            (self._mu3[iter] * self._W - self._rho)
+            + self._PsiT(self._mu2[iter] * self._U - self._eta)
+            + self._convolver.deconvolve(self._mu1[iter] * self._X - self._xi)
+        )
+
+        # rk = self._convolver._pad(rk)
+
+        freq_space_result = self._R_divmat[iter] * torch.fft.rfft2(rk, dim=(-3, -2))
+        self._image_est = torch.fft.irfft2(freq_space_result, dim=(-3, -2))
+
+        # self._image_est = self._convolver._crop(res)
+
+    def _W_update(self, iter):
+        """Non-negativity update"""
+        self._W = torch.maximum(
+            self._rho / self._mu3[iter] + self._image_est, torch.zeros_like(self._image_est)
+        )
+
+    def _xi_update(self, iter):
+        # to avoid computing forward model twice
+        self._xi = self._xi + self._mu1[iter] * (self._forward_out - self._X)
+
+    def _eta_update(self, iter):
+        # to avoid finite difference operataion again?
+        self._eta = self._eta + self._mu2[iter] * (self._Psi_out - self._U)
+
+    def _rho_update(self, iter):
+        self._rho = self._rho + self._mu3[iter] * (self._image_est - self._W)
+
+    def _update(self, iter):
+        self._U_update(iter)
+        self._X_update(iter)
+        self._W_update(iter)
+        self._image_update(iter)
+
+        # update forward and sparse operators
+        self._forward_out = self._convolver.convolve(self._image_est)
+        self._Psi_out = self._Psi(self._image_est)
+
+        self._xi_update(iter)
+        self._eta_update(iter)
+        self._rho_update(iter)
+
+    def _form_image(self):
+        image = self._convolver._crop(self._image_est)
+
+        # # TODO without cropping
+        # image = self._image_est
+
+        image[image < 0] = 0
+        return image
+
+    def batch_call(self, batch):
+        """
+        Method for performing iterative reconstruction on a batch of images.
+        This implementation is a properly vectorized implementation of ADMM.
+
+        Parameters
+        ----------
+        batch : :py:class:`~torch.Tensor` of shape (N, C, H, W) or (N, H, W, C)
+            The lensless images to reconstruct. If the shape is (N, C, H, W), the images are converted to (N, H, W, C) before reconstruction.
+
+        Returns
+        -------
+        :py:class:`~torch.Tensor` of shape (N, C, H, W) or (N, H, W, C)
+            The reconstructed images. Channels are in the same order as the input.
+        """
+        self._data = batch
+        batch_size = batch.shape[0]
+
+        if self._data.shape[-3] == 3:
+            CHW = True
```

Comment:
Using batch_call bypass set_data. We could assume this to be fix after the first call but this will make the function a bit less versatile (and not really any faster).

Reply:
can this be removed now that #41 is merged?